src/Dialect/Krnl/DialectBuilder.cpp

```diff
@@ -349,6 +349,49 @@ Value KrnlBuilder::constant(MemRefType type, StringRef name,
       alignment.value_or(nullptr));
 }
 
+//===----------------------------------------------------------------------===//
+// Math style functions.
+
+Value KrnlBuilder::roundEven(Value input) const {
+  Type elementType = getElementTypeOrSelf(input.getType());
+  MultiDialectBuilder<VectorBuilder, MathBuilder> create(*this);
+  //  hi alex, may want to generalize support to scalar as well.
+  VectorType vecType = mlir::dyn_cast<VectorType>(input.getType());
+  if (VectorMachineSupport::requireCustomASM(
+          GenericOps::roundEvenGop, elementType)) {
+    // Use Krnl round even op as LLVM does not support roundEven.
+    if (vecType) {
+      // Vector, enable unrolling of multiple archVL.
+      int64_t archVL = VectorMachineSupport::getArchVectorLength(
+          GenericOps::roundEvenGop, elementType);
+      assert(archVL > 1 && "expected vector with archVL>1");
+      assert(vecType.getRank() == 1 && "1D vec only");
+      int64_t vecSize = vecType.getShape()[0];
+      assert(vecSize % archVL == 0 && "expected multiple of archVL");
+      int64_t numArchVec = vecSize / archVL;
+      VectorType vecType2D = VectorType::get({numArchVec, archVL}, elementType);
+      Value input2D = create.vec.shapeCast(vecType2D, input);
+      Value output2D = input2D;
+      for (int64_t i = 0; i < numArchVec; ++i) {
+        Value subInput = create.vec.extractFrom2D(input2D, i);
+        Value subOutput =
+            b().create<KrnlRoundEvenOp>(loc(), subInput.getType(), subInput);
+        output2D = create.vec.insertInto2D(subOutput, output2D, i);
+      }
+      return create.vec.shapeCast(vecType, output2D);
+    } else {
```

Comment:
This `else` looks redundant since we have `return` inside the `if`.

Reply:
Tx, will fix